src/ip/QosConfig.cc

```diff
@@ -325,35 +326,30 @@ Ip::Qos::Config::parseConfigLine()
                        << "Netfilter mark preservation not available.");
 #endif // USE_LIBNETFILTERCONNTRACK
 #elif SO_MARK // SO_MARK && USE_LIBCAP
-                debugs(3, DBG_CRITICAL, "ERROR: Invalid parameter 'mark' in qos_flows option. "
-                       << "Linux Netfilter marking not available without LIBCAP support.");
-                self_destruct();
+                throw TextException(ToSBuf("Invalid parameter 'mark' in qos_flows option. ",
+                                           "Linux Netfilter marking not available on this platform."), Here());
 #else // SO_MARK && USE_LIBCAP
-                debugs(3, DBG_CRITICAL, "ERROR: Invalid parameter 'mark' in qos_flows option. "
-                       << "Linux Netfilter marking not available on this platform.");
-                self_destruct();
+                throw TextException(ToSBuf("Invalid parameter 'mark' in qos_flows option. ",
+                                           "Linux Netfilter marking not available on this platform."), Here());
 #endif // SO_MARK && USE_LIBCAP
             } else if (strncmp(token, "tos",3) == 0) {
                 preserveMissTos = true;
                 tos = true;
             } else {
-                preserveMissTos = true;
-                tos = true;
+                throw TextException(ToSBuf("Bad qos_flaws token value ", &token[0]), Here());
```

Comment:
AFAICT, there is [consensus](https://github.com/squid-cache/squid/pull/1644#issuecomment-1902855838) that we want to get rid of this "default to tos" behavior. Throwing an exception here is the right thing to do. However, throwing an exception is _not_ enough! The missing pieces include:

1. Update cf.data.pre to document the new/updated qos_flows configuration syntax. FWIW, without that update, I have no idea what syntax this PR is trying to support!
2. Move tos/mark determination away from the `while` loop (assuming the first tos/mark parameter is now required and must not be repeated).
3. Fix `qos_flaws` spelling in the new exception text.
4. Replace `token value` in the exception text with the name of the now-required qos_flows parameter from item 1 and suggest two of the supported values for that parameter (i.e. "tos" and "mark").
5. Update "Default to TOS for backwards compatibility" comment (currently at the start of the loop).
6. Update PR title/description to focus on this functionality change.
7. Update doc/release-notes/release-7.sgml.in to mention this functionality change.

FWIW, I recommend splitting this PR into two:

* A PR that replaces self_destruct() with exceptions. I recommend using this PR for that improvement. I have note checked carefully, but it is possible that all you need to do is to undo this `else` change.
* A PR that removes backward compatibility by updating this `else` clause and refactoring (see the seven items above). I recommend opening a new PR dedicated to that functionality change, after this PR merges.

Reply:
Good, will make as you suggested.
Could you check the rest changes?